_posts/2018-09-12-graphql-is-not-odata.md

```diff
@@ -0,0 +1,224 @@
+
+Another detail we have found important and highly successful: Our GraphQL layer IS NOT implemented or operated by the teams building RESTful services. The UI teams build that layer and Howard's team provides the platform and runs the service.
+
+This lets the service teams concentrate on REST and the domain model. GraphQL is an implementation detail of the UI layer—a technology chosen by UI, not services. This has avoided the whole REST vs. GraphQL debate with each of the dozens of service teams building APIs. They get to do their thing the way they want. For all they care, the UI consumes their services directly. We just happen to put a GraphQL server in between. We can centralize the GraphQL implementation into a smaller community of developers where we can foster reuse and commonalities more easily.
+
+Service teams are not burdened with *also* knowing and doing GraphQL. If they had to, they would inevitably look for ways to skip a layer, violating the separation of concerns. I often refer to using GraphQL as a having a double-paned window. UI cannot reach APIs directly; GraphQL is in between as insulation, reducing concerns on both sides. GraphQL helps both the UI and API layers as a result.
+
+### Downsides of GraphQL
+
+*[see in thread](https://twitter.com/JeffHandley/status/1035275129268236288)*
+
+Having to justify it when first adopting it or before enough service teams realize it will only benefit them.
+
+It is indeed yet another layer of ceremony where every single UI interaction needs to be thought through and implemented specifically (or covered by existing resolvers).
+
+It takes notable calendar time to design the schema thoughtfully such that it will have backwards compatibility as it grows over time. There is no versioning and the schema must be additive only (or with backwards compatible implementation detail changes). *Breaking changes are possible, but you must phase it: add new schema, rollout compensating changes to ALL clients, remove old schema.*
+
+It is yet another service to run and it is a single point of failure if not run well.
+
+Features like paging and filtering and sorting are not built in and require that you invest in defining your patterns and reusable types.
+
+For a big organization, you must intentionally create and foster a community of GraphQL contributors to ensure consistency and reuse. (Duplicating the same effort that should exist across API teams).
+
+It works best when there is a single GraphQL server fulfilling the whole schema. That requires a single service that multiple teams can contribute resolvers to. That is where Howard's team comes in for us—they build a platform we can publish resolver packages to.
+
+Coordination between Service teams, GraphQL teams, consuming teams and multiple staging/integration environments can be challenging.
+
+But I consider most of these to be human problems, not technology problems, and I like that trade-off.
```

Comment:
The one thing that I'll mention that is a downside of GraphQL, along with every other attempt at "strong typing" (from a serialization point of view) is that it doesn't do very well with dynamic data. The example in Concur's case is lists - even in GraphQL, we have to end up defining a wrapper data structure where the type information is conveyed along with the value information so that the client can do the right thing. And like I said, most of the RESTish attempts out there have this same problem.

Fortunately, there's that whole linked data thingy... :smile:

Reply:
Good point.  Glenn has also called out:

1. Cacheable queries don't work as well in GraphQL
2. Hypermedia also requires specific modeling; you can't just use content-type

Reply:
Originally, I held a similar view on hypermedia. However, the deeper I've gone into GraphQL, the more I've concluded that the localized benefits of hypermedia (e.g. local links to other services) aren't really necessary in GraphQL - where instead of links, I would use extensions and just let the client drill in whichever way she wished. 

For leveraging hypermedia as a way to bridge disparate systems, I would argue that a) this isn't really a goal of GraphQL in the first place, and b) aside from a formalized way of organizing links as a core part of the data model (e.g. linked data), you're going to rely on specific modeling for links anyway (whether that comes from you or whether it comes from the authors of HAL, Siren, Collection JSON, OData, etc.)

Reply:
Oh, interesting points.  I will remove the block I added that brings up the points Glenn's points since this itself could be a huge tangent that you and Glenn would be better poised to discuss than me.